Fix returning the proper last inserted id in MSSQL

sarven opened the pull request:
<!-- Fill in the relevant information below to help triage your pull request. -->

|      Q       |   A
|------------- | -----------
| Type         | bug
| BC Break     | yes
| Fixed issues | #3516 

#### Summary
I discovered a problem when the table in MSSQL has a trigger with insert statements. Id returned by method lastInsertId is invalid. It isn't id from the table in where we were inserting data, but from the table that trigger was inserting data.

The problem concerns sqlsrv and pdo_sqlsrv extensions. The fix for sqlsrv was quite simple, I only added a few lines of code. 

But the fix for pdo_sqlsrv required to rebuild the implementation of Connection and Statement. I made it similar to the sqlsrv implementation.

guilhermeblanco committed:
Decouple unique index from unique constraint

guilhermeblanco committed:
Added flag support to unique constraint

guilhermeblanco committed:
Fixed tests

guilhermeblanco committed:
Removed identifier generator type from schema table definition

guilhermeblanco committed:
Fixing rebase with master

guilhermeblanco committed:
Fixed undefined result of rebase

guilhermeblanco committed:
Moved getSequencePrefix() from ClassMetadata (ORM) to AbstractPlatform (DBAL)

morozov committed:
Updated UPGRADE.md for 3.x

morozov committed:
[BC] Removed support for PDO::PARAM_INPUT_OUTPUT in Statement::bindParam()

morozov committed:
[BC] Replaced PDO constants with values

morozov committed:
[BC] Replaced extension of \PDOStatement with a composition to avoid having to replicate the \PDOStatement interface in ResultStatement

Majkl578 committed:
Bump version to 3.0.x-dev

Majkl578 committed:
Drop Doctrine\DBAL\Version in favor of Ocramius\PackageVersions

BenMorel committed:
Drop Drizzle support

BenMorel committed:
Add note to UPGRADE.md

BenMorel committed:
Introduce a null SQL logger

BenMorel committed:
Make NullLogger final

BenMorel committed:
One-liner for Configuration::getSQLLogger()

BenMorel committed:
Change the SQLLogger interface signature

BenMorel committed:
Add UPGRADE note

BenMorel committed:
Ignore NullLogger in code coverage

BenMorel committed:
Make SQLLogger implementations final

BenMorel committed:
Update UPGRADE.md

BenMorel committed:
Do not accept null in Configuration::setSQLLogger()

BenMorel committed:
Add note about setSQLLogger()

BenMorel committed:
Don't catch Exception anymore

morozov committed:
When rendering SQL, only render the alias if it's different from the table name

Majkl578 committed:
Drop support for PostgreSQL <9.3

Majkl578 committed:
Drop support for SQL Server <2008

Majkl578 committed:
Drop support for SQL Anywhere <16

Majkl578 committed:
Drop PDO DB2 driver

Majkl578 committed:
Remove MsSQLKeywords

Majkl578 committed:
Reorder keywords and type mappings alphabetically

morozov committed:
Fixed build failure on SQL Server (sqlsrv)

The regression was introduced by 2b8c40d267b4ac699ea878dcca07332e1697978c.

Fixes #3060.

morozov committed:
Fixed build failure on SQL Server (pdo_sqlsrv)

Use a wrapped PDO Statement instead of a custom class for `pdo_sqlsrv` since `PDOSqlsrv\Statement` doesn't extend `PDOStatement` anymore.

morozov committed:
[DBAL-3079] Reworked the usage of PDO in PDOConnection from inheritance to composition

morozov committed:
[DBAL-3079] Added type hints to query-related method parameters and return values

morozov committed:
[DBAL-3079] Updated upgrade documentation

Majkl578 committed:
Revert complex type specification in Connection

Commit:
Added string mapping in SqlitePlatform - used by make bundle
  - fix 'Unknown database type string requested, Doctrine\DBAL\Platforms\SqlitePlatform may not support it.' error

Ocramius committed:
Merge pull request #3119 from DeyV/develop

Added "string" mapping in SqlitePlatform

Majkl578 committed:
Merge pull request #3113 from Majkl578/revert-complex-type-spec

Revert complex type specification in Connection

Majkl578 committed:
Extract exception factory methods into specific exceptions

Extract Doctrine\DBAL\DBALException::notSupported()
Extract Doctrine\DBAL\DBALException::invalidPlatformVersionSpecified()
Extract Doctrine\DBAL\DBALException::invalidPdoInstance()
Extract Doctrine\DBAL\DBALException::invalidWrapperClass()
Extract Doctrine\DBAL\DBALException::invalidDriverClass()
Extract Doctrine\DBAL\DBALException::invalidTableName()
Extract Doctrine\DBAL\DBALException::noColumnsSpecifiedForTable()
Extract Doctrine\DBAL\DBALException::typeExists()
Extract Doctrine\DBAL\DBALException::typeAlreadyRegistered()
Extract Doctrine\DBAL\DBALException::typeNotRegistered()
Extract Doctrine\DBAL\DBALException::unknownColumnType()
Extract Doctrine\DBAL\DBALException::typeNotFound()
Extract Doctrine\DBAL\DBALException::invalidPlatformType()
Extract Doctrine\DBAL\DBALException::driverRequired()
Extract Doctrine\DBAL\DBALException::unknownDriver()
Extract Doctrine\DBAL\DBALException::invalidColumnIndex()
Extract Doctrine\DBAL\Exception\ConnectionException::commitFailedRollbackOnly()
Extract Doctrine\DBAL\Exception\ConnectionException::noActiveTransaction()
Extract Doctrine\DBAL\Exception\ConnectionException::savepointsNotSupported()
Extract Doctrine\DBAL\Exception\ConnectionException::mayNotAlterNestedTransactionWithSavepointsInTransaction()
Extract Doctrine\DBAL\Exception\InvalidArgumentException::fromEmptyCriteria()
Extract Doctrine\DBAL\Type\ConversionException::conversionFailed()
Extract Doctrine\DBAL\Type\ConversionException::conversionFailedFormat()
Extract Doctrine\DBAL\Type\ConversionException::conversionFailedInvalidType()
Extract Doctrine\DBAL\Type\ConversionException::conversionFailedSerialization()
Extract Doctrine\DBAL\Query\QueryException::unknownAlias()
Extract Doctrine\DBAL\Query\QueryException::nonUniqueAlias()
Extract Doctrine\DBAL\Sharding\ShardingException::notImplemented()
Extract Doctrine\DBAL\Sharding\ShardingException::missingDefaultFederationName()
Extract Doctrine\DBAL\Sharding\ShardingException::missingDefaultDistributionKey()
Extract Doctrine\DBAL\Sharding\ShardingException::activeTransaction()
Extract Doctrine\DBAL\Sharding\ShardingException::noShardDistributionValue()
Extract Doctrine\DBAL\Sharding\ShardingException::missingDistributionType()
Extract Doctrine\DBAL\Cache\CacheException::noCacheKey()
Extract Doctrine\DBAL\Cache\CacheException::noResultDriverConfigured()
Extract Doctrine\DBAL\Schema\SchemaException::tableDoesNotExist()
Extract Doctrine\DBAL\Schema\SchemaException::indexNameInvalid()
Extract Doctrine\DBAL\Schema\SchemaException::indexDoesNotExist()
Extract Doctrine\DBAL\Schema\SchemaException::indexAlreadyExists()
Extract Doctrine\DBAL\Schema\SchemaException::columnDoesNotExist()
Extract Doctrine\DBAL\Schema\SchemaException::namespaceAlreadyExists()
Extract Doctrine\DBAL\Schema\SchemaException::tableAlreadyExists()
Extract Doctrine\DBAL\Schema\SchemaException::columnAlreadyExists()
Extract Doctrine\DBAL\Schema\SchemaException::sequenceAlreadyExists()
Extract Doctrine\DBAL\Schema\SchemaException::sequenceDoesNotExist()
Extract Doctrine\DBAL\Schema\SchemaException::uniqueConstraintDoesNotExist()
Extract Doctrine\DBAL\Schema\SchemaException::foreignKeyDoesNotExist()
Extract Doctrine\DBAL\Schema\SchemaException::namedForeignKeyRequired()
Extract Doctrine\DBAL\SQLParserUtilsException::missingParam()
Extract Doctrine\DBAL\SQLParserUtilsException::missingType()
Drop Doctrine\DBAL\DBALException::invalidPlatformSpecified()
Drop Doctrine\DBAL\DBALException::limitOffsetInvalid()
Drop Doctrine\DBAL\Schema\SchemaException::alterTableChangeNotSupported()

morozov committed:
Removed the logic of using BLOB columns for BINARY type of fields

morozov committed:
Removed support for DB-generated UUIDs

morozov committed:
Merge pull request #3211 from morozov/issues/3167

Removed support for DB-generated UUIDs

morozov committed:
Fixed and suppressed issues found by PHPStan on develop

Ocramius committed:
Merge pull request #3238 from morozov/phpstan-develop

Fixed and suppressed issues found by PHPStan on develop

morozov committed:
Removed dbal:import CLI command

morozov committed:
Merge pull request #3241 from morozov/issues/3237

Removed dbal:import CLI command

morozov committed:
Made the OFFSET in LIMIT queries non-nullable

morozov committed:
Merge pull request #3248 from morozov/non-nullable-offset

Made the OFFSET in LIMIT queries non-nullable integer defaulting to 0

Majkl578 committed:
Remove Doctrine\DBAL\Types\Type::getDefaultLength()

Majkl578 committed:
Remove Doctrine\DBAL\Types\Type::__toString()

morozov committed:
Got rid of func_num_args in AbstractPlatform::getDummySelectSQL()

morozov committed:
Handle binding errors in OCI8Statement::execute() and MySQLiStatement::execute()

morozov committed:
Statement::fetchColumn() will throw an exception in the case of invalid index

morozov committed:
Moved rowCount() from Statement to ResultStatement

Majkl578 committed:
Enable strict types

morozov committed:
Merge pull request #3417 from morozov/issues/3358

Statement::fetchColumn() will throw an exception in the case of invalid index

morozov committed:
Transaction-related Statement methods return void

morozov committed:
Connection::setAutoCommit() will return void

Ocramius committed:
Merge pull request #3480 from morozov/transaction-void

Transaction-related Statement methods return void

morozov committed:
Connection::ping() will return void

morozov committed:
Converted Connection and Statement methods which returned false in case of a failure into void

morozov committed:
Merge pull request #3483 from morozov/row-count

Moved rowCount() from Statement to ResultStatement

morozov committed:
Merge pull request #3486 from morozov/conn-stmt-void

Converted Connection and Statement methods which returned false in case of a failure into void

morozov committed:
Connection::quote() can only quote strings

morozov committed:
Assert that the username is specified on the connection when listing DB2 tables

morozov committed:
Make the $tableName argument of _getPortableTableIndexesList() required

1. The argument is always available since the method is only called from listTableIndexes() which requires a table name.
2. The argument itself seems a workaround and only needed to bypass the fact that the SQLiteSchemaManager violates the method contract: instead of reformatting the provided data it fetches more data from the DB schema which requires a table name. This argument should be dropped completely later.

Ocramius committed:
Merge pull request #3488 from morozov/quote-only-string

Connection::quote() can only quote strings

morozov committed:
Changed the type of `$char` in `AbstractPlatform::getTrimExpression()` from `string|false` to `?string`

Ocramius committed:
Merge pull request #3491 from morozov/get-trim-expr

Changed the type of $char in AbstractPlatform::getTrimExpression()`from string|false to ?string

morozov committed:
Modified AbstractPlatform::getLocateExpression() and ::getSubstringExpression() signatures

Ocramius committed:
Merge pull request #3494 from morozov/get-locate-expr

Modified AbstractPlatform::getLocateExpression() and ::getSubstringExpression() signatures

morozov committed:
Reworked AbstractPlatform::get*Expression() methods

Ocramius committed:
Merge pull request #3498 from morozov/get-other-expr

Reworked AbstractPlatform::get*Expression() methods

morozov committed:
Reworked `AbstractSchemaManager::extractDoctrineTypeFromComment()` removed `::removeDoctrineTypeFromComment()`

morozov committed:
Removed DB2SchemaManager::_getPortableForeignKeyRuleDef()

morozov committed:
Merge pull request #2854 from Majkl578/strict-types

Enable strict types

Majkl578 committed:
Merge pull request #3502 from morozov/db2-get-portable-foreign-key-rule-def

Removed DB2SchemaManager::_getPortableForeignKeyRuleDef()

morozov committed:
Reworking driver exceptions

Ocramius committed:
Merge pull request #3505 from morozov/driver-exceptions

Reworked driver exceptions

morozov committed:
Removed errorCode() and errorInfo() methods from Connection and Statement APIs

Ocramius committed:
Merge pull request #3507 from morozov/error-code-info

Removed errorCode() and errorInfo() methods from Connection and Statement APIs

morozov committed:
Added type hints to the Column class

morozov committed:
Merge pull request #3501 from morozov/extract-type-from-comment

Reworked AbstractSchemaManager::extractDoctrineTypeFromComment(), removed ::removeDoctrineTypeFromComment()

morozov committed:
Implemented support for persistent connections in PDO and mysqli drivers

Majkl578 committed:
Remove deprecated json_array type

Majkl578 committed:
Remove deprecated Connection::TRANSACTION_* constants

Majkl578 committed:
Remove deprecated AbstractPlatform::DATE_INTERVAL_UNIT_* and AbstractPlatform::TRIM_* constants

Majkl578 committed:
Remove deprecated MysqlSessionInit listener

Majkl578 committed:
Remove deprecated MysqlPlatform::getCollationFieldDeclaration()

Majkl578 committed:
Remove deprecated AbstractPlatform::getIdentityColumnNullInsertSQL()

Majkl578 committed:
Remove deprecated Table::addUnnamedForeignKeyConstraint() and Table::addNamedForeignKeyConstraint()

Majkl578 committed:
SQLParserUtils::*_TOKEN constants made private

Majkl578 committed:
Remove deprecated SQLParserUtils::getPlaceholderPositions()

sarven committed:
add test for issue 3516 - invalid last inserted id in mssql when the table has a trigger with insert

Majkl578 committed:
Remove deprecated Table::renameColumn()

Majkl578 committed:
Remove deprecated AbstractSchemaManager::getFilterSchemaAssetsExpression(), Configuration::getFilterSchemaAssetsExpression() and Configuration::getFilterSchemaAssetsExpression()

sarven committed:
fix getting proper last inserted id when the table has a trigger with insert stmt in mssql

sarven committed:
change the number of insert statements in the trigger

jwage committed:
Fix UniqueConstraint with empty name.

jwage committed:
Drop support for feferred misspelling in
PostgreSqlPlatform::getAdvancedForeignKeyOptionsSQL()

jwage committed:
Merge pull request #3520 from doctrine/drop-misspelling-in-foreign-key-constraints

Drop support for feferred misspelling in PostgreSqlPlatform::getAdvancedForeignKeyOptionsSQL()

morozov committed:
Merge pull request #3511 from morozov/column-type-hints

Added type hints to the Column class

sarven committed:
fix PDOSqlsrv to get proper last inserted id when the table has a trigger with an insert statement

sarven committed:
fix phpstan errors

sarven committed:
fix phpcs

sarven committed:
fix phpdoc for stmt

sarven committed:
fix typehint

sarven committed:
convert fetchColumn result to string

sarven committed:
remove unnecessary if

sarven committed:
remove whitespace

sarven committed:
convert fetchColumn to string

sarven committed:
remove getAttribute method which is not in the Connection interface

sarven committed:
fix typehint

jwage committed:
Merge pull request #3519 from doctrine/fix-unique-constraint-with-empty-name

Fix UniqueConstraint with empty name.

morozov committed:
Merge pull request #3515 from morozov/persistent-connections

Implemented support for persistent connections in PDO and mysqli drivers

morozov committed:
Merge pull request #3525 from doctrine/exceptions

Extract exception factory methods into specific exceptions

jwage committed:
Improve consistency of exception message formatting.

morozov committed:
Merge pull request #3536 from jwage/improve-exception-messages

Improve consistency of exception message formatting.

sarven committed:
change sqlsrv version from devel to stable

SenseException commented on tests/Doctrine/Tests/DBAL/Functional/Driver/PDOSqlsrv/DriverTest.php:
```diff
@@ -64,11 +63,4 @@ public function testConnectionOptions() : void
         self::assertSame('APP_NAME', $result);
     }
-
-    public function testDriverOptions() : void
```
Is this test breaking with your changes?

sarven replied:
Yes, because previously Connection extended PDOConnection, now it only implements Connection interface, where is no method like getAttribute.

SenseException replied:
master-branch is currently for 2.10 while develop is for 3.0. It looks like the changes are more fitting for the next major release.

sarven replied:
So should I make a branch based on develop then apply these changes and create a new pull request or just override this?

sarven committed:
adjust changes to develop branch

morozov committed:
Remove user provided PDO functionality

morozov committed:
Merge pull request #3548 from morozov/issues/3545

Remove user provided PDO functionality

Ocramius committed:
Merge pull request #3485 from morozov/ping-void

`Connection::ping()` will now return `void`

morozov committed:
Enforced parameter and return value types in Statement classes

morozov committed:
Merge pull request #3552 from morozov/statement-types

Enforced parameter and return value types in Statement classes

morozov commented:
@sarven could you rebase this please?

morozov committed:
Merge pull request #3518 from Majkl578/drop-deprecated

Remove @​deprecated stuff

morozov committed:
Removed Driver::getName()

The method is not used for anything else than skipping tests for specific drivers. Cross-driver portability should be established by drivers, not outside of them based on their name.

sarven commented:
@morozov Done.

sarven committed:
adjust PDOSqlsrvStatement

morozov commented on .appveyor.yml:
```diff
@@ -68,7 +68,7 @@ install:
           Add-Content php.ini "`n curl.cainfo=C:\tools\cacert\bundle.pem"
 
-          $DLLVersion = "5.5.0preview"
+          $DLLVersion = "5.6.1"
```
Please revert unless this is required. If possible, we should roll back to using the latest stable version as it was prior to #3484 but that would be a different PR.

sarven replied:
@morozov I reverted, but probably some tests won't pass.

sarven replied:
@morozv Now in appveyor is the error:
`
PHP Fatal error:  Meta data vector out of sync in
`
Should I create a new pull request with a change of version?

morozov commented on lib/Doctrine/DBAL/Driver/PDOSqlsrv/Statement.php:
```diff
@@ -1,37 +0,0 @@
-/**
- * PDO SQL Server Statement
- */
-class Statement extends PDOStatement
```
Why does this need to be renamed?

sarven replied:
I just wanted to name it similarly to SQLSrvStatement. It is not subtype of PDOStatement, so I wanted to mark it clearly.

morozov reviewed (commented):
@sarven thank you. Could you please describe the problem in the pull request description?

morozov commented on tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL3516Test.php:
```diff
@@ -0,0 +1,46 @@
+
+        $platform = $this->connection->getDatabasePlatform()->getName();
+
+        if ($platform !== 'mssql') {
```
Since this is not the only driver/platform which implements `lastInsertId()`, the test should be implemented for all supported platforms and skipped if there's a known issue with a specific platform.

sarven replied:
I thought about make this tests for all platforms, but there is a problem with create trigger for all platforms. So I think that creating it only for MSSQL is ok, because this problem probably doesn't concern others platforms.

morozov commented on tests/Doctrine/Tests/DBAL/Functional/Ticket/DBAL3516Test.php:
```diff
@@ -0,0 +1,46 @@
+/**
+ * @group DBAL-3516
+ */
+class DBAL3516Test extends DbalFunctionalTestCase
```
Please move it under `Doctrine\Tests\DBAL\Functional\ConnectionTest`.

sarven replied:
Ok

morozov commented on tests/Doctrine/Tests/DBAL/Functional/StatementTest.php:
```diff
@@ -373,7 +374,9 @@ public function testExecWithRedundantParameters() : void
-        if ($this->connection->getWrappedConnection() instanceof PDOConnection) {
+        $wrappedConnection = $this->connection->getWrappedConnection();
+
+        if ($wrappedConnection instanceof PDOConnection || $wrappedConnection instanceof PDOSqlsrvConnection) {
```
Why isn't `PDOSqlsrvConnection` sub-type of `PDOConnection`?

sarven replied:
I did some changes and now Connection and Statement in PDOSqlsrv are subtypes of specific doctrine PDO classes. Previously I wanted it to be similarly to the Sqlsrv implementation, but it is also possible to just extends PDO classes and do some overrides. Now it seems to be simpler.

morozov committed:
Merge pull request #3553 from morozov/remove-driver-name

Removed Driver::getName()

sarven committed:
Revert "change sqlsrv version from devel to stable"

This reverts commit 7efabf04dc6f4c069b7cfe60eea68b74ad1cf85a.

sarven committed:
move DBAL3516Test to ConnectionTest

sarven committed:
Connection in PDOSqlsrv as subtype of PDOConnection

sarven committed:
fix alias

Ocramius commented on lib/Doctrine/DBAL/Driver/PDOSqlsrv/Connection.php:
```diff
@@ -14,19 +17,31 @@
+    /**
+     * Append to any INSERT query to retrieve the last insert id.
+     */
+    protected const LAST_INSERT_ID_SQL = ';SELECT SCOPE_IDENTITY() AS LastInsertId;';
```
`private` unless needed elsewhere

Ocramius commented on lib/Doctrine/DBAL/Driver/PDOSqlsrv/Connection.php:
```diff
@@ -14,19 +17,31 @@
 class Connection extends PDOConnection
 {
+    /** @var LastInsertId|null */
+    protected $lastInsertId;
```
`private` unless needed elsewhere

Ocramius commented on lib/Doctrine/DBAL/Driver/PDOSqlsrv/LastInsertId.php:
```diff
@@ -0,0 +1,30 @@
+ */
+class LastInsertId
+{
+    /** @var string */
```
According to (missing) constructor semantics, this is `string|null`

Ocramius commented on lib/Doctrine/DBAL/Driver/PDOSqlsrv/LastInsertId.php:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    /**
+     * @return string
```
According to (missing) constructor semantics, this is `string|null`

Ocramius commented on lib/Doctrine/DBAL/Driver/PDOSqlsrv/LastInsertId.php:
```diff
@@ -0,0 +1,30 @@
+    /**
+     * @param string $id
+     */
+    public function setId($id)
```
Type declarations over docblocks: this is `string` and `: void`. Same for the rest of the patch overall.

Ocramius reviewed (changes requested):
I'd need @deeky666 to check this patch, since he initially authored https://github.com/doctrine/dbal/pull/2648 (which was closed due to its high complexity/risk for little gain).

I'd say that also here the same applies: if a prepared statement is causing side-effects through stored procedures and triggers, I'd rather call the schema "foobar'd" and "needs to be fixed there" rather than trying to fix issues on this end.

sarven committed:
change protected to private

sarven committed:
add missing typehints to LastInsertId

sarven committed:
fix cs

sarven committed:
id as nullable

warslett commented:
hey @sarven @Ocramius any update on this PR? looks like the branches might need clearing up (presuming this PR doesn't include 141 commits).
